File: array_adt/exercises/multipleMissingElements.cc
// find the missing elements in an sorted array

#include <iostream>
using namespace std;

void missingValues(int array[], int difference, int length){
    // time complexity is O(n)
    for(int i = 0; i < length; i++){
        while (difference < (array[i]-i)){
            cout << "Missing number is " << (i+difference) << endl;
            difference += 1;
        }
    }
}

// print missing elements in one line
void missingValues2(int array[], int difference, int length){
    int elements[4], index=  0;

    for(int i = 0; i < length; i++){
        while (difference < (array[i]-i)){
            elements[index] = (i+difference);
            difference += 1;
            index++;
        }
    }

    cout << "Missing elements are ";
    for(auto & i: elements){
        cout << i << " ";
    }
    cout << endl;
}

int main(){
    int array[] = {6,7,8,9,11,12,15,16,17,18,19};
    int difference = array[0] - 0; // firstvalue - index
    int length = sizeof(array)/sizeof(int);

    missingValues(array, difference, length);
    missingValues2(array, difference, length);
}

File: array_adt/exercises/multipleMissingElements.py
# find the multiple missing numbers in an sorted array

# input Array elements
# find difference between currentArrayValue and value's index
# while i < arrayLength
# while difference < currentElementDifference (if missing elements are more than 1 at a time)
# then print the missing number (difference + currentIndex)
# increment the difference


print("Enter the elements", end=' ')
array = list(map(int, input().split()))
difference = array[0] - 0 # firstValue - index

# time is O(n)
for i in range(len(array)):
    while (difference < array[i]-i):
        print(f'Missing value is {i+difference}')
        difference += 1

# print missing values in one line
diff = array[0] - 0
missingValues = list()

for i in range(len(array)):
    while (diff < array[i]-i):
        missingValues.append((i+diff))
        diff += 1

print(f'The missing values are', end=' ')
for i in missingValues:
    print(i, end=' ')

print()